aiidalab_widgets_base/viewers.py

```diff
@@ -766,68 +1083,111 @@ class StructureDataViewer(_StructureDataBaseViewer):
     """
 
     structure = tl.Union(
-        [tl.Instance(ase.Atoms), tl.Instance(orm.Node)], allow_none=True
+        [
+            tl.Instance(ase.Atoms),
+            tl.Instance(orm.StructureData),
+            tl.Instance(orm.CifData),
+        ],
+        allow_none=True,
     )
     displayed_structure = tl.Instance(ase.Atoms, allow_none=True, read_only=True)
     pk = tl.Int(allow_none=True)
 
     def __init__(self, structure=None, **kwargs):
         super().__init__(**kwargs)
         self.structure = structure
-        self.natom = len(self.structure) if self.structure is not None else 0
 
     @tl.observe("supercell")
-    def repeat(self, _=None):
+    def _observe_supercell(self, _=None):
         if self.structure is not None:
+            self.set_trait(
+                "displayed_structure", None
+            )  # To make sure the structure is always updated.
             self.set_trait("displayed_structure", self.structure.repeat(self.supercell))
 
     @tl.validate("structure")
-    def _valid_structure(self, change):  # pylint: disable=no-self-use
+    def _valid_structure(self, change):
         """Update structure."""
         structure = change["value"]
-
-        if structure is None:
-            return None  # if no structure provided, the rest of the code can be skipped
-
         if isinstance(structure, ase.Atoms):
             self.pk = None
-            return structure
-        if isinstance(structure, orm.Node):
+        elif isinstance(structure, (orm.StructureData, orm.CifData)):
             self.pk = structure.pk
-            return structure.get_ase()
-        raise TypeError(
```

Comment:
Why was this removed? Because traitlets take care of the validation already?

Reply:
Yes, exactly for that. The code never reaches this point.